src/utils/json.ts

```diff
@@ -0,0 +1,23 @@
+import { JsonSafe } from '@stencil/core/declarations';
+
+/**
+ * A thin wrapper around JSON.stringify which ensures that the value to be
+ * stringified is a valid JSON-serializable value, as codified by the {@link
+ * JsonSafe} type. If you try to serialize a type which isn't valid you'll get
+ * a type error.
+ *
+ * **NOTE**: This is _only_ a type-level safety guarantee. No runtime check is
+ * made at all that the passed value has a serializable runtime type.
+ *
+ * @param toStringify a value to be serialized
+ * @param replacer an optional replacer (same as JSON.stringify)
+ * @param space an  optional space to use (same as JSON.stringify)
+ * @returns a serialized string
+ */
+export function safeJSONStringify<T>(
```

Comment:
we could change this name, I do think this type-level assurance is a helpful thing to add but wouldn't want to convey more safety than there is here (i.e. as the JSDoc notes there is no runtime check at all, it's just a type-level check)